### PR TITLE
bug(schema): do not write network-config when instance dir absent

### PR DESCRIPTION
**Pull request by blackboxsw**
## Proposed Commit Message
```
bug(schema): write network-config if instance dir present (#4635)

Only write /var/lib/cloud/instance/network-config.json once
datasource is detected. The /var/lib/clound/instance symlink is
created by Init.instancify after datasource detection is complete.

Move creation of /var/lib/cloud/instance/network-config.json into
it's own method _write_network_config_json. It will be called by
any call to apply_network_config.

apply_network_config is called in both Local and Network stages.

In Local stage, apply_network_config is used to either:
 - render the final network config of datasource detected in Local
 - in absence of Local datasource, render basic fallback DHCP on
   primary NIC to allow network to come up before detecting a
   Network datasource

For Network datasources, they will not have been discovered or
instancify'd in Local boot stage, so apply_network_config cannot
yet persist network-config.json.

Defer creation of network-config.json for Network datasources
until until the link /var/lib/cloud/instance exists and
apply_network_config is called in Network stage to render final
network config.

Fixes GH-4630
```

## Additional Context
This broke live-service autoinstall images in Noble which use DataSourceNone for configuration on first boot because Local boot stage tried to create /var/lib/cloud/instance/tmpasdffile when atomically writing network-config.json to a directory /var/lib/cloud/instance which didn't exist yet.

## Test Steps

```
# reproducer 
lxc launch ubuntu-daily/noble test-n
lxc exec test-n -- apt update
lxc exec test-n -- apt install cloud-init
lxc exec test-n -- cloud-init --version  | grep 23.4
# inject datasource None config
cat > 99-none.cfg <<EOF
datasource_list: [None]
datasource:
  None:
    metadata: 
       instance-id: my-id
    userdata_raw: "#cloud-config\nssh_import_id: [YOUR_LP_NAME]"
EOF
lxc file push 99-none.cfg test-n/etc/cloud/cloud.cfg.d/
lxc exec test-n -- cloud-init clean --logs --reboot
lxc exec test-n -- cloud-init status --wait --format=json
# expect to see Tracebacks w/ FileNotFound related to writing /var/lib/cloud/instance/*
```

```
# build local deb from this PR
./tools/run-container --package ubuntu/noble
lxc file push cloud-init*bddeb.deb test-n/
lxc exec test-n -- apt install /cloud-init*deb
lxc exec test-n -- cloud-init clean --logs --reboot
lxc exec test-n -- cloud-init status --wait --format=json
# confirm /var/lib/cloud/instance/network-config.json still exists for the Network detected datasource
```

## Checklist

- [x] My code follows the process laid out in [the documentation](https://cloudinit.readthedocs.io/en/latest/development/index.html)
- [x] I have updated or added any [unit tests](https://cloudinit.readthedocs.io/en/latest/development/testing.html) accordingly
- [ ] I have updated or added any [documentation](https://cloudinit.readthedocs.io/en/latest/development/contribute_docs.html) accordingly

## Merge type

- [ ] Squash merge using "Proposed Commit Message"
- [x] Rebase and merge unique commits. Requires commit messages per-commit each referencing the pull request number (#<PR_NUM>)

### CODE REVIEW

**Comment by TheRealFalcon** [comment]
@blackboxsw do we know the root cause here? When was this introduced?

**Comment by blackboxsw** [comment]
> 4630

@TheRealFalcon it's this commit ee86a37fefed39f53fe140926e66dcd19f8c87dc. The reason we don't see it in our integration tests is because the primary datasources we validate are all init Local timeframe https://github.com/canonical/cloud-init/issues/4630#issuecomment-1830359302. I can setup a DataSourceNone one off integration test on LXD that at least will assert cloud-init doesn't see errors/warnings for a very basic config If we want. That'd give us a signal if something else breaks in the future.

**Comment by TheRealFalcon** [comment]
> I can setup a DataSourceNone one off integration test on LXD that at least will assert cloud-init doesn't see errors/warnings for a very basic config If we want. That'd give us a signal if something else breaks in the future.

Lets do that. If not now, we should at least create an issue for it

**Comment by holmanb** [comment]
@blackboxsw Prior to ee86a37fefed39f53fe140926e66dcd19f8c87dc, what `_apply_netcfg_names()` did was pretty obvious. It was a thin wrapper around `self.distro.networking.apply_network_config_names(netcfg)`.

After that commit, it includes a collection of looking up paths and conditionally reading and writing files. Reading it, it's not at all clear what it is supposed to be doing and why. We need to add some context this function in comments at a minimum, but I'm not convinced that this function is the right place for this code to live. It seems unrelated to the interface renaming that happens in that function. Why should it be there?

Note that `_apply_netcfg_names()` is called twice - once before the call to `validate_cloudconfig_schema()`, and once after. The change in ee86a37fefed39f53fe1 was to add schema validation to network config - why do we write the network config again after validating? What is the point?

It may be that this is the right place for this code to live for now, and that this is the right fix. However, given my questions above, I think we need to justify in the comments why we're doing what we are, since it really isn't obvious.

**Review by holmanb — CHANGES_REQUESTED** [review]
Request for some additional comments at a minimum, and possibly reconsider where this code lives.

File: cloudinit/stages.py [hunk 1]
```diff
@@ -403,6 +404,24 @@ def _reflect_cur_instance(self):
 
         # Write out information on what is being used for the current instance
         # and what may have been used for a previous instance...
+        if getattr(self, "_netcfg"):
```
Comment (TheRealFalcon) [inline cloudinit/stages.py:407]: Nit: since we've already initialized it, `if self._netcfg:` should be fine and a little clearer.
Nit2: This function is getting a bit long. Since this behavior is already self-contained, does it make sense as a separate function call?

File: cloudinit/stages.py [hunk 2]
```diff
@@ -934,22 +953,7 @@ def _find_networking_config(self):
...
-                    atomic_helper.write_json(
-                        ncfg_instance_path, netcfg, mode=0o600
-                    )
-            else:
-                atomic_helper.write_json(
-                    ncfg_instance_path, netcfg, mode=0o600
-                )
-        if not os.path.islink(network_link):
-            util.sym_link(ncfg_instance_path, network_link)
+        self._netcfg = netcfg
```
Comment (TheRealFalcon) [inline cloudinit/stages.py:956]: I don't think we should set this here this as we have always first called `self._find_networking_config()` before we call this.

**Review by TheRealFalcon — CHANGES_REQUESTED** [review]
Haven't reviewed tests yet, but otherwise my comments are inline.

File: tests/integration_tests/datasources/test_none.py [hunk 3]
```diff
@@ -0,0 +1,64 @@
+"""NoCloud datasource integration tests."""
```
Comment (TheRealFalcon) [inline tests/integration_tests/datasources/test_none.py:1]: This isn't NoCloud

File: tests/integration_tests/datasources/test_none.py [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+    Test that DataSourceNone detection (used by live installers) doesn't
+    generate errors or warnings.
+    """
+    log = client.read_from_file("/var/log/cloud-init.log")
+    verify_clean_log(log)
+    # Limit datasource detection to DataSourceNone.
+    client.write_to_file(
+        "/etc/cloud/cloud.cfg.d/99-force-dsnone.cfg", DS_NONE_BASE_CFG
+    )
+    client.execute("cloud-init clean --logs --reboot")
```
Comment (TheRealFalcon) [inline tests/integration_tests/datasources/test_none.py:58]: I think we should do `client.execute("cloud-init clean --logs")` followed by `client.restart()` here, otherwise there's a possibility your `wait_for_cloud_init_status_file()` will return before the instance has even gone down.

File: tests/integration_tests/datasources/test_none.py [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+"""
+
+
+@retry(tries=30, delay=1)
+def wait_for_cloud_init_status_file(instance: LXDInstance):
+    status_file = instance.read_from_file("/run/cloud-init/status.json")
+    assert len(status_file)
+
+
+@pytest.mark.lxd_use_exec
```
Comment (TheRealFalcon) [inline tests/integration_tests/datasources/test_none.py:41]: Why is this necessary?

File: tests/integration_tests/datasources/test_none.py [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+@pytest.mark.skipif(
+    PLATFORM != "lxd_container",
+    reason="Requires LXD container with custom metadata setup",
```
Comment (TheRealFalcon) [inline tests/integration_tests/datasources/test_none.py:44]: Is this still true? Can we remove this skip?

File: tests/integration_tests/datasources/test_none.py [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+  None:
+    metadata:
+      instance-id: my-iid-uuid
+    userdata_raw: |
+      #cloud-config
+      runcmd:
+      - touch /var/tmp/success-with-datasource-none
+"""
+
+LXD_METADATA_DS_NONE_BASE_CFG = """\
```
Comment (TheRealFalcon) [inline tests/integration_tests/datasources/test_none.py:24]: I don't see this used anywhere

File: tests/integration_tests/datasources/test_none.py [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+    client.write_to_file(
+        "/etc/cloud/cloud.cfg.d/99-force-dsnone.cfg", DS_NONE_BASE_CFG
+    )
+    client.execute("cloud-init clean --logs --reboot")
+    wait_for_cloud_init_status_file(client)
+    status = json.loads(client.execute("cloud-init status --format=json"))
+    assert [] == status["errors"]
+    assert {} == status["recoverable_errors"]
+    client.execute("cloud-init status --wait")
+    assert client.execute("test -f /var/tmp/success-with-datasource-none").ok
```
Comment (TheRealFalcon) [inline tests/integration_tests/datasources/test_none.py:64]: Is it worth another `verify_clean_log` call after the reboot?

**Review by TheRealFalcon — CHANGES_REQUESTED** [review]
The rest of my comments

File: cloudinit/stages.py (hunk 2 above)
Comment (blackboxsw) [reply cloudinit/stages.py:956]: right, dropped. It's already being set on the instance.

File: tests/integration_tests/datasources/test_none.py (hunk 6 above)
Comment (blackboxsw) [reply tests/integration_tests/datasources/test_none.py:44]: yes we can drop this now that we are manipulating a running instance which already has networking.

File: cloudinit/stages.py (hunk 1 above)
Comment (blackboxsw) [reply cloudinit/stages.py]: You are correct here, But, I'm dropping this as approach and going back to my original because `apply_network_config` is called in 3 circumstances:
 1. In Local stage after Local datasource is detected (instancify already called at ds detection)
 2. In Local stage before Network datasource is detected to render basic net config on primary NIC (instancify not yet called)
 3. In Network stage after a Network datasource is detected (instancify already called at ds detection)

In all cases, apply_network_config will not have been called prior to instancify, so no preserved _net_cfg is yet set.

File: cloudinit/stages.py (hunk 2 above)
Comment (blackboxsw) [reply cloudinit/stages.py]: dropped in latest. We will perform this in apply_network_config with smarts to avoid writing the file unless `/var/lib/cloud/instance` exists.

File: tests/integration_tests/datasources/test_none.py (hunk 4 above)
Comment (blackboxsw) [reply tests/integration_tests/datasources/test_none.py:61]: `wait_for_cloud_init_status_file` doesn't return unless /run/cloud-init/status.json is non-empty. We won't get to this case because cloud-init clean trashes the link target in `/v/l/c/instance/` that `/run/cloud-init/status.json` points to so any attempt to cat /run/cloud-init/status.json returns FileNotFound errors which exits non-zero. Our `read_from_file` raises an IOError in those cases which triggers a retry on `wait_for_cloud_init_status_file`.  The reason I want to avoid client.restart() is because it blocks until cloud-init completes, and in the failure mode here, cloud-init never completes due to unhandled Tracebacks which leave `cloud-init status --wait` blocking in a running state indefinitely. Since `client.restart()` in our integration tests doesn't provide a `wait=False` I constructed the wait_for_cloud_init_status_file as a workaround for this case to at least give us visibility to any warning/error conditions that show up even if cloud-init never completes.

File: tests/integration_tests/datasources/test_none.py (hunk 6 above)
Comment (blackboxsw) [reply tests/integration_tests/datasources/test_none.py]: There is also a known bug with fallback network config on lxd_containers with virtual NICs https://bugs.launchpad.net/netplan/+bug/2022947 that requires that we provide network config that doesn't emit a "match:` clause by mac address due to netplan 0.106 behavior. For this case, we can inject the known working network `version: 2` configuration to make sure we can still ssh into lxd_containers across the DataSourceNone boundary (which generates fallback network config).

File: tests/integration_tests/datasources/test_none.py (hunk 8 above)
Comment (blackboxsw) [reply tests/integration_tests/datasources/test_none.py:85]: @TheRealFalcon sure. I've adapted `verify_clean_log` to append known warnings when detected DataSourceNone. We will expect to have both "Used fallback datasource"  and occasional "Cannot call Open vSwitch" dbus race.

File: tests/integration_tests/datasources/test_none.py (hunk 4 above)
Comment (TheRealFalcon) [reply tests/integration_tests/datasources/test_none.py:61]: Ah, right. I wasn't thinking about clean removing the link. This is good

**Review by TheRealFalcon — APPROVED** [review]
LGTM...but let's wait for @holmanb too

**Comment by blackboxsw** [comment]
> @blackboxsw The integration test fails locally for me:
> 
> ```
> ================================================ FAILURES ================================================
> _____________________________________ test_datasource_none_discovery _____________________________________
> 
> client = <tests.integration_tests.instances.IntegrationInstance object at 0x7f2f22c65950>
> 
>     def test_datasource_none_discovery(client: IntegrationInstance):
>         """Integration test for #4635.
>     
>         Test that DataSourceNone detection (used by live installers) doesn't
>         generate errors or warnings.
>         """
>         log = client.read_from_file("/var/log/cloud-init.log")
>         verify_clean_log(log)
>         # Limit datasource detection to DataSourceNone.
>         client.write_to_file(
>             "/etc/cloud/cloud.cfg.d/99-force-dsnone.cfg", DS_NONE_BASE_CFG
>         )
>         if client.settings.PLATFORM in ["lxd_container"]:
>             # DataSourceNone provides no network_config.
>             # To avoid changing network config from platform desired net cfg
>             # to fallback config, copy out the rendered network config
>             # to /etc/cloud/cloud.cfg.d/99-orig-net.cfg so it is
>             # setup by the DataSourceNone case as well.
>             # Otherwise (LXD specifically) we'll have network torn down due
>             # to virtual NICs present which results in not network being
>             # brought up when we emit fallback config which attempts to
>             # match on PermanentMACAddress. LP:#2022947
>             client.execute(
>                 "cp /etc/netplan/50-cloud-init.yaml"
>                 " /etc/cloud/cloud.cfg.d/99-orig-net.cfg"
>             )
>         client.execute("cloud-init clean --logs --reboot")
>         wait_for_cloud_init_status_file(client)
>         status = json.loads(client.execute("cloud-init status --format=json"))
>         assert [] == status["errors"]
>         expected_warnings = set(
>             [
>                 "Used fallback datasource",
>                 "Running ['netplan', 'apply'] resulted in stderr output: Cannot"
>                 " call Open vSwitch: ovsdb-server.service is not running.\nFailed"
>                 " to connect system bus: No such file or directory\nFalling back"
>                 " to a hard restart of systemd-networkd.service\n",
>             ]
>         )
>         unexpected_warnings = set(
>             status["recoverable_errors"].get("WARNING", [])
>         ).difference(expected_warnings)
>         if unexpected_warnings:
> >           raise AssertionError(
>                 f"Unexpected recoverable errors: {list(unexpected_warnings)}"
>             )
> E           AssertionError: Unexpected recoverable errors: ["Running ['netplan', 'apply'] resulted in stderr output: \x1b[0;1;31mFailed to connect system bus: No such file or directory\x1b[0m\nFalling back to a hard restart of systemd-networkd.service\n"]
> 
> tests/integration_tests/datasources/test_none.py:78: AssertionError
> ----------------------------------------- Captured stdout setup ------------------------------------------
> Settings:
> CLOUD_INIT_SOURCE=IN_PLACE
> COLLECT_LOGS=ON_ERROR
> EXISTING_INSTANCE_ID=None
> INSTANCE_TYPE=None
> KEEP_IMAGE=False
> KEEP_INSTANCE=False
> LOCAL_LOG_PATH=/tmp/cloud_init_test_logs
> OS_IMAGE=focal
> PLATFORM=lxd_container
> RUN_UNSTABLE=False
> Launching instance with launch_kwargs:
> image_id=ubuntu-daily:94b416961b506181b6704e7620f835a66d495f0290c598386518441e0b87d55c
> user_data=None
> username=ubuntu
> Mounting source /home/holmanb/cloud-init-g/cloudinit directly onto LXD container/VM named cloudinit-1130-181514gvs928my at /usr/lib/python3/dist-packages/cloudinit
> Mounting source /home/holmanb/cloud-init-g/templates directly onto LXD container/VM named cloudinit-1130-181514gvs928my at /etc/cloud/templates
> Launched instance: LXDInstance(name=cloudinit-1130-181514gvs928my)
> cloud-init version: /usr/bin/cloud-init 23.3.3
> image serial: 20231129
> ------------------------------------------- Captured log setup -------------------------------------------
> 2023-11-30 11:15:14 INFO      integration_testing:clouds.py:76 Settings:
> CLOUD_INIT_SOURCE=IN_PLACE
> COLLECT_LOGS=ON_ERROR
> EXISTING_INSTANCE_ID=None
> INSTANCE_TYPE=None
> KEEP_IMAGE=False
> KEEP_INSTANCE=False
> LOCAL_LOG_PATH=/tmp/cloud_init_test_logs
> OS_IMAGE=focal
> PLATFORM=lxd_container
> RUN_UNSTABLE=False
> 2023-11-30 11:15:14 INFO      integration_testing:clouds.py:143 Launching instance with launch_kwargs:
> image_id=ubuntu-daily:94b416961b506181b6704e7620f835a66d495f0290c598386518441e0b87d55c
> user_data=None
> username=ubuntu
> 2023-11-30 11:15:15 INFO      pycloudlib.cloud.LXDContainer:cloud.py:252 ['lxc', 'init', 'ubuntu-daily:94b416961b506181b6704e7620f835a66d495f0290c598386518441e0b87d55c', 'cloudinit-1130-181514gvs928my', '--config', 'user.meta-data=public-keys: ssh-rsa AAAAB3NzaC1yc2EAAAADAQABAAACAQCqpCozNyoIR4yR5UGrjUg4HzfDs0voCchtGzRcYc1HO83rgPJGHMSFi5WuiLCZtWMtLShnZBxwouXo03mnYPe1dC1QBas2QZIEgJ9Piq+c3kS/YxIf0R2UNg/laYeRNpNtz5zuC+gBQpz5BWLGL39PUHyW6eBtIdpl2CrOASN56pK6Hk285ZphFVWDc+GyxqM1s1E1qH8NmtX2RKl9/LV/LWcPqwkwWMaV9PmdoiwKekJb9xNJiE0HEwdgE55FCSVzMMxJtGZaSHfva8qDamW/IK5JB7RM5u2aC71yWZI8DOXEPHFxKBWBJO4eMQPLQp4ACq8fEWqzci3Ieu9AzmGBa4PCiWA70flPoU09U8daqoYjZei/az1H0CAObaWeDo/pU8bN4CUX8ap3+RU3UzD5JMcZs9jy2VG1nbrM5CYM9EO4Mt6yT+/vQQbXd9Qvg3KkLM8IUVNUf0yPRb/yqJ3QknQ11LVdhJI8GPFzxC5JxbaqngAoqqE5Da2lHb+60vVXIJdwPgKQCXnD9E40Ej8XwmE3DJciZN8zFl1eNsAQF/wJIdoR7+uKnayMu6DcoOBwNwc2MjPJochkicLz7slpzslAE78XM2AzTZg5z8t4bAN+IWeOdiqlyvBFOVEBuzyVmhbkedvzj/SKciU8RFHup6STEH++ujOgIkBdqJYLPQ== holmanb@arc\n']
> 2023-11-30 11:15:21 INFO      integration_testing:clouds.py:307 Mounting source /home/holmanb/cloud-init-g/cloudinit directly onto LXD container/VM named cloudinit-1130-181514gvs928my at /usr/lib/python3/dist-packages/cloudinit
> 2023-11-30 11:15:21 INFO      integration_testing:clouds.py:307 Mounting source /home/holmanb/cloud-init-g/templates directly onto LXD container/VM named cloudinit-1130-181514gvs928my at /etc/cloud/templates
> 2023-11-30 11:15:22 INFO      pycloudlib.instance:instance.py:494 _wait_for_execute to complete
> 2023-11-30 11:15:22 WARNING   pycloudlib.instance:instance.py:431 The specified key (/home/holmanb/.ssh/id_rsa) requires a passphrase. If you have not added this key to a running SSH agent, you will see failures to connect after a long timeout.
> 2023-11-30 11:15:26 WARNING   pycloudlib.instance:instance.py:431 The specified key (/home/holmanb/.ssh/id_rsa) requires a passphrase. If you have not added this key to a running SSH agent, you will see failures to connect after a long timeout.
> 2023-11-30 11:15:27 WARNING   pycloudlib.instance:instance.py:431 The specified key (/home/holmanb/.ssh/id_rsa) requires a passphrase. If you have not added this key to a running SSH agent, you will see failures to connect after a long timeout.
> 2023-11-30 11:15:28 WARNING   pycloudlib.instance:instance.py:431 The specified key (/home/holmanb/.ssh/id_rsa) requires a passphrase. If you have not added this key to a running SSH agent, you will see failures to connect after a long timeout.
> 2023-11-30 11:15:28 INFO      paramiko.transport:transport.py:1893 Connected (version 2.0, client OpenSSH_8.2p1)
> 2023-11-30 11:15:28 INFO      paramiko.transport:transport.py:1893 Authentication (publickey) successful!
> 2023-11-30 11:15:29 INFO      pycloudlib.instance:instance.py:515 _wait_for_cloudinit to complete
> 2023-11-30 11:15:29 INFO      pycloudlib.instance:instance.py:270 executing: sh -c 'command -v systemctl'
> 2023-11-30 11:15:36 INFO      pycloudlib.instance:instance.py:270 executing: cloud-init status --wait --long
> 2023-11-30 11:15:37 INFO      integration_testing:clouds.py:154 Launched instance: LXDInstance(name=cloudinit-1130-181514gvs928my)
> 2023-11-30 11:15:37 INFO      pycloudlib.instance:instance.py:270 executing: sudo -- sh -c 'cloud-init --version'
> 2023-11-30 11:15:37 INFO      integration_testing:clouds.py:158 cloud-init version: /usr/bin/cloud-init 23.3.3
> 2023-11-30 11:15:37 INFO      pycloudlib.instance:instance.py:270 executing: sudo -- sh -c 'grep serial /etc/cloud/build.info'
> 2023-11-30 11:15:37 INFO      integration_testing:clouds.py:164 image serial: 20231129
> ------------------------------------------- Captured log call --------------------------------------------
> 2023-11-30 11:15:37 INFO      pycloudlib.instance:instance.py:270 executing: sudo -- sh -c 'cat /var/log/cloud-init.log'
> 2023-11-30 11:15:37 INFO      paramiko.transport.sftp:sftp.py:169 [chan 14] Opened sftp connection (server version 3)
> 2023-11-30 11:15:37 INFO      pycloudlib.instance:instance.py:270 executing: sudo -- sh -c 'mv /var/tmp/7430db19-9589-4f8f-afe5-be30bc4365a5.tmp /etc/cloud/cloud.cfg.d/99-force-dsnone.cfg'
> 2023-11-30 11:15:37 INFO      pycloudlib.instance:instance.py:270 executing: sudo -- sh -c 'cp /etc/netplan/50-cloud-init.yaml /etc/cloud/cloud.cfg.d/99-orig-net.cfg'
> 2023-11-30 11:15:37 INFO      pycloudlib.instance:instance.py:270 executing: sudo -- sh -c 'cloud-init clean --logs --reboot'
> 2023-11-30 11:15:38 INFO      pycloudlib.instance:instance.py:270 executing: sudo -- sh -c 'cat /run/cloud-init/status.json'
> 2023-11-30 11:15:38 WARNING   pycloudlib.instance:instance.py:431 The specified key (/home/holmanb/.ssh/id_rsa) requires a passphrase. If you have not added this key to a running SSH agent, you will see failures to connect after a long timeout.
> 2023-11-30 11:15:39 INFO      pycloudlib.instance:instance.py:270 executing: sudo -- sh -c 'cat /run/cloud-init/status.json'
> 2023-11-30 11:15:39 WARNING   pycloudlib.instance:instance.py:431 The specified key (/home/holmanb/.ssh/id_rsa) requires a passphrase. If you have not added this key to a running SSH agent, you will see failures to connect after a long timeout.
> 2023-11-30 11:15:43 INFO      pycloudlib.instance:instance.py:270 executing: sudo -- sh -c 'cat /run/cloud-init/status.json'
> 2023-11-30 11:15:43 WARNING   pycloudlib.instance:instance.py:431 The specified key (/home/holmanb/.ssh/id_rsa) requires a passphrase. If you have not added this key to a running SSH agent, you will see failures to connect after a long timeout.
> 2023-11-30 11:15:44 INFO      pycloudlib.instance:instance.py:270 executing: sudo -- sh -c 'cat /run/cloud-init/status.json'
> 2023-11-30 11:15:44 WARNING   pycloudlib.instance:instance.py:431 The specified key (/home/holmanb/.ssh/id_rsa) requires a passphrase. If you have not added this key to a running SSH agent, you will see failures to connect after a long timeout.
> 2023-11-30 11:15:45 INFO      pycloudlib.instance:instance.py:270 executing: sudo -- sh -c 'cat /run/cloud-init/status.json'
> 2023-11-30 11:15:45 WARNING   pycloudlib.instance:instance.py:431 The specified key (/home/holmanb/.ssh/id_rsa) requires a passphrase. If you have not added this key to a running SSH agent, you will see failures to connect after a long timeout.
> 2023-11-30 11:15:45 INFO      paramiko.transport:transport.py:1893 Connected (version 2.0, client OpenSSH_8.2p1)
> 2023-11-30 11:15:45 INFO      paramiko.transport:transport.py:1893 Authentication (publickey) successful!
> 2023-11-30 11:15:46 INFO      pycloudlib.instance:instance.py:270 executing: sudo -- sh -c 'cloud-init status --format=json'
> ---------------------------------------- Captured stdout teardown ----------------------------------------
> Writing logs to /tmp/cloud_init_test_logs/231130111512/tests/integration_tests/datasources/test_none/test_datasource_none_discovery
> ----------------------------------------- Captured log teardown ------------------------------------------
> 2023-11-30 11:15:47 INFO      pycloudlib.instance:instance.py:270 executing: sudo -- sh -c 'cloud-init collect-logs -u -t /var/tmp/cloud-init.tar.gz'
> 2023-11-30 11:15:48 INFO      integration_testing:conftest.py:169 Writing logs to /tmp/cloud_init_test_logs/231130111512/tests/integration_tests/datasources/test_none/test_datasource_none_discovery
> 2023-11-30 11:15:48 INFO      pycloudlib.instance:instance.py:270 executing: sh -c 'cp /var/tmp/cloud-init.tar.gz /var/tmp/ecd79a90-369b-408e-906f-cfb260382be0.tmp'
> 2023-11-30 11:15:48 INFO      paramiko.transport.sftp:sftp.py:169 [chan 4] Opened sftp connection (server version 3)
> ============================================ warnings summary ============================================
> .tox/integration-tests/lib/python3.11/site-packages/_pytest/config/__init__.py:1316
>   /home/holmanb/cloud-init-g/.tox/integration-tests/lib/python3.11/site-packages/_pytest/config/__init__.py:1316: PytestRemovedIn8Warning: The --strict option is deprecated, use --strict-markers instead.
>     self.issue_config_time_warning(
> 
> .tox/integration-tests/lib/python3.11/site-packages/pyparsing.py:108
>   /home/holmanb/cloud-init-g/.tox/integration-tests/lib/python3.11/site-packages/pyparsing.py:108: DeprecationWarning: module 'sre_constants' is deprecated
>     import sre_constants
> 
> -- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
> ========================================== slowest 10 durations ==========================================
> 25.60s setup    tests/integration_tests/datasources/test_none.py::test_datasource_none_discovery
> 9.63s call     tests/integration_tests/datasources/test_none.py::test_datasource_none_discovery
> 3.85s teardown tests/integration_tests/datasources/test_none.py::test_datasource_none_discovery
> ======================================== short test summary info =========================================
> FAILED tests/integration_tests/datasources/test_none.py::test_datasource_none_discovery - AssertionError: Unexpected recoverable errors: ["Running ['netplan', 'apply'] resulted in stderr outp...
> ===================================== 1 failed, 2 warnings in 39.27s =====================================
> ERROR: InvocationError for command /home/holmanb/cloud-init-g/.tox/integration-tests/bin/python -m pytest -vv --log-cli-level=INFO --durations 10 tests/integration_tests/datasources/test_none.py (exited with code 1)
> ________________________________________________ summary _________________________________________________
> ERROR:   integration-tests: commands failed
> ```

@holmanb we confirmed it's different messaging from netplan on older series. I've adapted the test (and verify_clean_log) to be match known unique substrings of those logs. Also tests/integration-tests/util.py was broadly matching "WARN" instead of cloud-init's formatted "[WARNING]"  levels which hit invalid hits for some of netplans formatted logging. Now this integration test works for F, J, L M and N series.

**Review by holmanb — APPROVED** [review]
The line lengths on the proposed commit message is long, but I trust that you can fix on merge.

LGTM, thanks!
